src/udev/memory_id/dmidecode.c

```diff
@@ -0,0 +1,768 @@
+}
+
+/*
+ * 7.17 Physical Memory Array (Type 16)
+ */
+
+static const char *dmi_memory_array_location(uint8_t code)
+{
+        /* 7.17.1 */
+        static const char *location[] = {
+                "Other", /* 0x01 */
+                "Unknown",
+                "System Board Or Motherboard",
+                "ISA Add-on Card",
+                "EISA Add-on Card",
+                "PCI Add-on Card",
+                "MCA Add-on Card",
+                "PCMCIA Add-on Card",
+                "Proprietary Add-on Card",
+                "NuBus" /* 0x0A */
+        };
+        static const char *location_0xA0[] = {
+                "PC-98/C20 Add-on Card", /* 0xA0 */
+                "PC-98/C24 Add-on Card",
+                "PC-98/E Add-on Card",
+                "PC-98/Local Bus Add-on Card",
+                "CXL Flexbus 1.0" /* 0xA4 */
+        };
+
+        if (code >= 0x01 && code <= 0x0A)
+                return location[code - 0x01];
+        if (code >= 0xA0 && code <= 0xA4)
+                return location_0xA0[code - 0xA0];
+        return out_of_spec;
+}
+
+static const char *dmi_memory_array_ec_type(uint8_t code)
+{
+        /* 7.17.3 */
+        static const char *type[] = {
+                "Other", /* 0x01 */
+                "Unknown",
+                "None",
+                "Parity",
+                "Single-bit ECC",
+                "Multi-bit ECC",
+                "CRC" /* 0x07 */
+        };
+
+        if (code >= 0x01 && code <= 0x07)
+                return type[code - 0x01];
+        return out_of_spec;
+}
+
+/*
+ * 7.18 Memory Device (Type 17)
+ */
+
+static void dmi_memory_device_width(const char *attr_prefix, const char *attr_suffix, unsigned int slot_num, uint16_t code)
+{
+        /*
+         * If no memory module is present, width may be 0
+         */
+        if (code != 0xFFFF && code != 0)
+                printf("%s_%u_%s=%u\n", attr_prefix, slot_num, attr_suffix, code);
+}
+
+static void dmi_memory_device_size(unsigned int slot_num, uint16_t code)
+{
+        if (code == 0) {
+                printf("MEMORY_DEVICE_%u_PRESENT=0\n", slot_num);
+                return;
+        } else if (code == 0xFFFF) {
+                return;
+        } else {
+                u64 s = { .l = code & 0x7FFF };
+                if (!(code & 0x8000))
+                        s.l <<= 10;
+                dmi_print_memory_size("MEMORY_DEVICE", "SIZE", slot_num, s, 1);
+        }
+}
+
+static void dmi_memory_device_extended_size(unsigned int slot_num, uint32_t code)
+{
+        uint64_t capacity;
+
+        capacity = code * 1024 * 1024;
+
+        printf("MEMORY_DEVICE_%u_SIZE=%"PRIu64"\n", slot_num, capacity);
+}
+
+static void dmi_memory_voltage_value(const char *attr_prefix, const char *attr_suffix, unsigned int slot_num, uint16_t code)
+{
+        if (code == 0)
+                return;
+        if (code % 100)
+                printf("%s_%u_%s=%g\n", attr_prefix, slot_num, attr_suffix, (float)code / 1000);
```

Comment:
"float" is such a weird type. We should never use that unless in an array where the memory savings matter. for all other cases use "double" please. In paticular as "%g" actually outputs a double, not a float anyway.